File: Colaslistas/colalis.cpp
#include <iostream>

class Nodo {
public:
    int data;
    Nodo* next;

    Nodo(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class ColaLista {
private:
    Nodo* front;
    Nodo* rear;

public:
    ColaLista() {
        this->front = nullptr;
        this->rear = nullptr;
    }

    bool isEmpty() {
        return this->front == nullptr;
    }

    void enqueue(int data) {
        Nodo* nuevoNodo = new Nodo(data);
        if (this->rear == nullptr) {
            this->front = nuevoNodo;
            this->rear = nuevoNodo;
            return;
        }
        this->rear->next = nuevoNodo;
        this->rear = nuevoNodo;
    }

    int dequeue() {
        if (this->isEmpty()) {
            std::cout << "Error: Cola subdesbordada (Underflow)" << std::endl;
            return -1;
        }

        int data = this->front->data;
        Nodo* temp = this->front;
        this->front = this->front->next;

        if (this->front == nullptr) {
            this->rear = nullptr;
        }

        delete temp;
        return data;
    }

    int peek() {
        if (this->isEmpty()) {
            std::cout << "Cola vacía" << std::endl;
            return -1;
        }
        return this->front->data;
    }
};

int main() {
    ColaLista cola;
    cola.enqueue(10);
    cola.enqueue(20);
    cola.enqueue(30);

    std::cout << "Elemento frontal: " << cola.peek() << std::endl;
    std::cout << "Elimina elemento: " << cola.dequeue() << std::endl;
    std::cout << "Nuevo elemento frontal: " << cola.peek() << std::endl;

    cola.enqueue(40);
    std::cout << "Elimina elemento: " << cola.dequeue() << std::endl;
    std::cout << "Elimina elemento: " << cola.dequeue() << std::endl;
    std::cout << "Elimina elemento: " << cola.dequeue() << std::endl;
    std::cout << "Elimina elemento: " << cola.dequeue() << std::endl;

    return 0;
}

File: Colaslistas/colalis.js
class Nodo {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class ColaLista {
    constructor() {
        this.front = null;
        this.rear = null;
    }

    isEmpty() {
        return this.front === null;
    }

    enqueue(data) {
        const nuevoNodo = new Nodo(data);
        if (this.rear === null) {
            this.front = nuevoNodo;
            this.rear = nuevoNodo;
            return;
        }
        this.rear.next = nuevoNodo;
        this.rear = nuevoNodo;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("Error: Cola subdesbordada (Underflow)");
            return null;
        }

        const data = this.front.data;
        this.front = this.front.next;

        if (this.front === null) {
            this.rear = null;
        }
        return data;
    }

    peek() {
        if (this.isEmpty()) {
            console.log("Cola vacía");
            return null;
        }
        return this.front.data;
    }
}

const cola = new ColaLista();
cola.enqueue(10);
cola.enqueue(20);
cola.enqueue(30);

console.log(`Elemento frontal: ${cola.peek()}`);
console.log(`Elimina elemento: ${cola.dequeue()}`);
console.log(`Nuevo elemento frontal: ${cola.peek()}`);

cola.enqueue(40);
console.log(`Elimina elemento: ${cola.dequeue()}`);
console.log(`Elimina elemento: ${cola.dequeue()}`);
console.log(`Elimina elemento: ${cola.dequeue()}`);
console.log(`Elimina elemento: ${cola.dequeue()}`);